Implementation of Two Factor Authentication

MarcelBolten opened the pull request:
Hi Nicolas,

I started to work on a first implementation of two factor authentication #2075.
- It uses the PHP library for Two Factor Authentication and the QR code is generated by the mPDF QR code library.
- 2FA is only applied to 'local' logins. It is not used for SAML or external authentication by web server.
- The users will get a database entry that holds the 2FA secret/token or it is null.
- Only users will be able to activate 2FA but sys(admins) can disable it in case users lock themselves out.
- If 2FA is activated there will be an auto redirect to request the verification code. A failed attempt is counted as an invalid login and users get banned based on the already implemented anti brute force system.

I tried to make it as modular as possible but it can probably be optimized.

Cheers,
Marcel

NicolasCARPi commented on src/classes/App.php:
```diff
@@ -75,7 +76,11 @@ public function __construct(Request $request, Session $session, Config $config,
         $this->Session = $session;
-        $this->ok = $this->Session->getFlashBag()->get('ok');
+
+        // Don't get 'ok' falshes during mfa
```
typo

MarcelBolten replied:
Fixed

NicolasCARPi commented on src/classes/Auth.php:
```diff
@@ -134,6 +144,8 @@ public function needAuth(): bool
             'login.php',
             'LoginController.php',
             'metadata.php',
+            'mfa.php',
```
we should avoid adding more single php files. maybe use login.php and LoginController instead

MarcelBolten replied:
My idea of having the MFA code in a separate files is that one could use MFA not only during login but maybe also somewhere else. E.g. if user changes the password or if (sys)admin does something delicate or maybe set a timeout after which one needs to reverify.
Why do you want to avoid the addition of more files?

NicolasCARPi replied:
Because php files that end up in the URL means more maintenance, and less clean code. Look at experiments.php and database.php, they are nearly the same, and yet the code lives in different files. That's because of "history". But here we don't need to do the same mistake. Having the code tucked away in a controller/class doesn't stop one from using it anywhere (which is a good idea btw). I would really like to avoid having more direct access php files in web/.

MarcelBolten replied:
Less maintenance makes sense to me. I transplanted the logic into login.php and LoginController.php, as suggested.
For me it works very smoothly now. However, I did not test whether it interferes with SAML or authentication by web server.

NicolasCARPi commented on src/classes/Mfa.php:
```diff
@@ -0,0 +1,178 @@
+     */
+    public function __construct(Request $request, Session $session)
+    {
+        $this->TwoFactorAuth = new TwoFactorAuth('eLabFTW', 6, 30, 'sha1', new MpdfQrProvider());
```
let's use sha256 instead. Everybody is moving away from sha1

MarcelBolten replied:
[It seems to be a problem](https://labanskoller.se/blog/2019/07/11/many-common-mobile-authenticator-apps-accept-qr-codes-for-modes-they-dont-support/) to use something else than the 'default' (HMAC-SHA-1 algorithm, 6 digits and a period of 30 seconds) if you want to have a broad range of authenticator apps working. There is a [recommendation to use default values](https://labanskoller.se/blog/2019/07/11/many-common-mobile-authenticator-apps-accept-qr-codes-for-modes-they-dont-support/#recommendation-to-site-owners). In addition, the token does not need to be extremely secure as a potential attacker also needs the password and will be banned after some tries due to the anti brute force system.

On the other hand, eLabFTW could recommend an app that works with sha256. However, users might try to use an app they already have installed anyways and potentially will complain about 2FA not working on eLabFTW.

NicolasCARPi replied:
Ok, let's keep sha1 then!

NicolasCARPi commented on src/classes/Mfa.php:
```diff
@@ -0,0 +1,178 @@
+    }
+
+    /**
+     * Load MFA secret of user from databse if exists
```
typo

MarcelBolten replied:
Fixed

NicolasCARPi commented on src/classes/Mfa.php:
```diff
@@ -0,0 +1,178 @@
+     * Load MFA secret of user from databse if exists
+     *
+     * @param int $userid
+     * @return mixed MFA secret or false
```
maybe throw exception instead of returning false. We should avoid mixed returns

MarcelBolten replied:
Now it returns the secret or an empty string. No more mixed return here.

NicolasCARPi commented on src/classes/Mfa.php:
```diff
@@ -0,0 +1,178 @@
+
+        $Response = new RedirectResponse('../../mfa.php');
+        $Response->send();
+        exit;
```
there should never be an exit in a class method, return the Response and let client code deal with it.

MarcelBolten replied:
exit is removed

NicolasCARPi commented on src/classes/Mfa.php:
```diff
@@ -0,0 +1,178 @@
+
+        $Response = new RedirectResponse($location);
+        $Response->send();
+        exit;
```
same as previous

MarcelBolten replied:
exit is removed

NicolasCARPi commented on src/models/Users.php:
```diff
@@ -362,6 +364,14 @@ public function update(array $params): void
+        if ($params['use_mfa'] === 'off') {
+            $mfaSql = ', mfa_secret = null';
+        } elseif ($params['use_mfa'] === 'on' && !$this->userData['mfa_secret']) {
+            throw new ImproperActionException('Only users themselves can activate two factor authetication!');
```
typo

MarcelBolten replied:
Fixed

NicolasCARPi commented:
Also, use `yarn phpcs`, `yarn psalm` and `yarn phpstan` and check your code doesn't add new issues ;)

NicolasCARPi commented on src/classes/App.php:
```diff
@@ -79,7 +79,11 @@ public function __construct(Request $request, Session $session, Config $config,
         $this->Session = $session;
-        $this->ok = $this->Session->getFlashBag()->get('ok');
+
+        // Don't get 'ok' flashes during mfa
```
why is this needed? wouldn't it be better to make sure nothing is sent to the ok bag during mfa? (it's always better to avoid ifs)

MarcelBolten replied:
A user needs to enable 2FA in the _Account_ tab of the _User Control Panel_. At the same time other settings can be changed. Before this PR, this changes are confirmed with a 'Saved' message. However, with 2FA enabled there will be a few redirects (UcpController.php -> login.php -> LoginController.php -> ucp.php?tab=2). A User will get the secret and can setup their app, eLab will verify the 2FA code and enable 2FA. On the way many things can go wrong but the other setting are stored first. Therefore, the 'Saved' message is added to the flashbag directly (open to discussion).
I see several options here:
1. Show the 'Saved' message on the login.php page. **Pro**: no need to do the if. **Con**: Maybe it is confusing for the user to get the message here.
2. Don't add the 'Saved' message at all if user also wants to enable 2FA. **Pro**: no need to do the if. **Con**: User gets no feedback that the other settings are stored. But can see it later in the UCP.
3. Only add the 'Saved' message after 2FA was successfully enabled. **Pro**: no need to do the if. **Con**: User only gets feedback if all works as intended but no feedback in any other case.
4. Maybe it is possible to use a Session variable? I would need to think about it a bit. **Pro**: no need to do the if. **Con**: ?

NicolasCARPi replied:
I'll need to run your code and see by myself how it can be done.

NicolasCARPi commented on src/templates/ucp.html:
```diff
@@ -2,10 +2,10 @@
-        <li class='tabhandle' id='tab2'>{{ 'Account'|trans }}</li>
-        <li class='tabhandle' id='tab3'>{{ 'Templates'|trans }}</li>
-        <li class='tabhandle' id='tab4'>{{ 'API keys'|trans }}</li>
+  <li class='tabhandle' id='tab2'>{{ 'Account'|trans }}</li>
```
why is this inverted?

MarcelBolten replied:
Good question, must be by mistake. I fixed it.

MarcelBolten commented:
I removed stuff not related to 2FA. The indentation was also reverted to match the hypernext branch. It should make it easier to follow the changes.
I can help to fix the indentation in a different PR if wished.

NicolasCARPi commented:
@MarcelBolten Right now I'm focusing on fixing the bugs in 3.5, and once this is out of the way I'll try your code so it can be merged in 3.6. Thanks for your patience.

NicolasCARPi commented on src/classes/Mfa.php:
```diff
@@ -0,0 +1,199 @@
+    public function abortEnable(): string
+    {
+        $this->Session->getFlashBag()->add('ko', _('Two Factor Authentication not enabled!'));
+        $location = $this->Session->get('mfa_redirect');
```
`$location` is not used here. Found through: https://scrutinizer-ci.com/g/elabftw/elabftw/inspections/ab841c9a-62cd-4b66-84ce-ff1f13a3349d/issues/files/src/classes/Mfa.php?status=new&orderField=path&order=asc&honorSelectedPaths=0

MarcelBolten replied:
This is a leftover. I removed it. MFA::cleanup() handles the redirect location now.

NicolasCARPi commented:
I started running the code. What is the best workflow to test this out extensively? Meaning can I somehow emulate a phone with the app installed? What are the supported apps? My phone is so old I'm not even sure I could install such apps; I need a way to debug and play with it in dev mode.

NicolasCARPi commented:
I'm thinking it would be nice if you could open a PR on the next branch of elabdoc and write a few lines on 2FA auth in the User guide.

MarcelBolten commented:
I think what needs to be tested is the set up in the UCP account tab. This is the entry point for a user to use 2FA.
Next would be to logout and login with 2FA enabled.
Last thing is to disable 2FA from the account tab or as (sys)admin users tab.

Ideally I can provide some unit tests but I don't know how the test environment works. I didn't look into it so far.

The PHP library for Two Factor Authentication can be used to emulate a phone. This will require a few lines of code. What feature do you need for testing? I used the apps listed in https://github.com/elabftw/elabftw/issues/2075#issue-671593374 successfully but any app should work.

I will add some lines to elabdoc and open a PR.

MarcelBolten commented:
> Meaning can I somehow emulate a phone with the app installed?

I added a very simple PHP command line tool (src/tools/2fa_tools.php) which will return a 2FA code if you provide the secret as a first argument. It can be used like the tools in the bin folder. I.e.: `php src/tools/2fa_tools.php 2FA:code EXAMPLE2FASECRET234567ABCDEFGHIJ` The secret will be provided during the set up of 2FA from the UCP account tab and you will want to save it somewhere for testing. The tool could be modified to get the secret straight from the database.

> My phone is so old I'm not even sure I could install such apps

I think that should be no problem unless you are still using the good old Nokia 3310.

NicolasCARPi reviewed (commented):
nice, I'll test that out asap. Currently I have to work on a performance issue :/

MarcelBolten commented:
I added some unit tests.

NicolasCARPi commented on src/tools/2fa_tools.php:
```diff
@@ -0,0 +1,78 @@
+            '2FA code: ' . (string) $code,
+        ));
+
+        return 1;
```
return 0 if nothing went wrong

MarcelBolten replied:
Changed with https://github.com/elabftw/elabftw/pull/2076/commits/367e560c6e05da7a429da93f63afee6dad0fb18e

NicolasCARPi commented:
![2020-09-09-170956_1094x442_scrot](https://user-images.githubusercontent.com/3043706/92617376-83ac3d80-f2bf-11ea-8ec2-248b81ebd8b2.png)

MarcelBolten commented:
PopulateDatabase.php (https://github.com/elabftw/elabftw/pull/2076#issuecomment-689627503) fixed with https://github.com/elabftw/elabftw/pull/2076/commits/73e51e5f371866d03a8681097942032d1bbbc4ad

NicolasCARPi commented:
I tested your code. It seems to work fine :+1: 

The 2fa_tools is handy but if it's in the Elabftw\Commands namespace it should live in src/commands

And we could call it with `bin/console dev:2fa`.

MarcelBolten commented:
> The 2fa_tools is handy but if it's in the Elabftw\Commands namespace it should live in src/commands
> And we could call it with `bin/console dev:2fa`.

I changed the location of the 2fa_tools to src/commands. It can now be called with `bin/console dev:2fa`. I also renamed the file and class to MfaCode. This will avoid the class name duplication scrutinizer was complaining about (https://github.com/elabftw/elabftw/pull/2076#issuecomment-689627503) and makes the use of an alias obsolete.

I might have broken the acceptance tests by adding a 2FA secret for the first user in the populate-config.yml. In particular the part in _bootstrap.php might not work anymore. But I don't have a pipeline running to test it myself.

NicolasCARPi reviewed (approved):
I'm sure I'll change a few things here and there afterwards, but the code is green, working, so I see no reason not to merge this! :)

Thanks again for this very good work you did there! :clinking_glasses:

NicolasCARPi commented:
> I might have broken the acceptance tests

Can confirm. Fixed by just setting this to false, unit tests for 2fa still pass. I'll come back to it anyway at some point before 3.6.

>  But I don't have a pipeline running to test it myself.

What happens if you try `yarn test` ? If you are in docker group it should work imho.

MarcelBolten commented:
It might be a bit cumbersome but I run ubuntu without docker in a hyper-v virtual machine on my windows computer. Works for most of the stuff but obviously `yarn test` will not.

NicolasCARPi commented:
Time to get a real OS ;)
